File: code/matlab/problem3a.m
% Problem 3(a) – Worst-Case Sidelobe Minimization (20 antennas)
% -------------------------------------------------------------------------
% Rather than average sidelobe energy, we now bound the worst sidelobe
% magnitude over the stop band.  CVX variable t is the upper envelope.

clear all;
clc;
close all;

% for part 3(a), set NUM_ANA=20
NUM_ANA = 20;  % Number of antennas, directly controls array aperture.

% setup
CARR_FREQ = 2.4e9;                         % Hz. carrier frequency.
WAVE_LEN = physconst("LightSpeed") / CARR_FREQ;  % Lambda = c / f.
ANTENNA_DIS = WAVE_LEN/2;                  % Spacing (half lambda).

% beam width (from theta_l to theta_u)
theta_l = 10;
theta_u = 30;

% beam center (at the middle of theta_l to theta_u)
ANGLE_DES = (theta_l + theta_u)/2;  % Desired steering angle.

% Beam steering vector a(theta)
Steering_des = exp(-1j*(0:NUM_ANA-1).' * 2*pi*ANTENNA_DIS*sin(ANGLE_DES*pi/180)/WAVE_LEN);


% Minimize worst-case sidelobe --------------------------------------
P_matrix = zeros(NUM_ANA, NUM_ANA);  % Placeholder reused in the loop.

cvx_begin
    variable t                     % Worst sidelobe energy (epigraph var).
    variable w(NUM_ANA,1) complex   % Beamformer to be designed.
    expression match_output
    minimize(t);
    subject to
        % Loop over sidelobe angles and force the quadratic form <= t.
        for i=-90:1:theta_l
            P_matrix = exp(-1j*(0:NUM_ANA-1)'*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN) * ...
                       exp(-1j*(0:NUM_ANA-1)'*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN)';
            match_output = quad_form(w, P_matrix);
            match_output <= t;
        end
        for i=theta_u:1:90
            P_matrix = exp(-1j*(0:NUM_ANA-1)'*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN) * ...
                       exp(-1j*(0:NUM_ANA-1)'*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN)';
            match_output = quad_form(w, P_matrix);
            match_output <= t;
        end

        % Preserve unity gain towards the desired look direction.
        w'*Steering_des == 1;
cvx_end

% plot angle spectrum
steering_vec_plot = [];
for i=-90:1:90
    steering_vec_plot = [steering_vec_plot; exp(-1j*(0:NUM_ANA-1)*2*pi*ANTENNA_DIS*sin(i*pi/180)/WAVE_LEN)];
end

% plot
f = figure;
plot(-90:1:90, 10*log10(abs(w'*steering_vec_plot.').^2));
title('Minimizing the Worst-case Sidelobe')

hold on
grid on
xlabel('Angle(degree)');
ylabel('Angle Response(db)');
xlim([-90, 90])
%ylim([-120, 0])
hold off

saveas(f, 'problem3a.png')
savefig(f, 'problem3a.fig')

File: code/python/problem3a.py
"""Worst-case sidelobe minimization for a 20-element ULA."""

import pickle

import numpy as np
import cvxpy as cp
import matplotlib.pyplot as plt


NUM_ANA = 20
CARR_FREQ = 2.4e9
LIGHT_SPEED = 299_792_458.0
WAVE_LEN = LIGHT_SPEED / CARR_FREQ
ANTENNA_DIS = WAVE_LEN / 2

theta_l = 10
theta_u = 30
ANGLE_DES = (theta_l + theta_u) / 2  # target look direction

indices = np.arange(NUM_ANA)
steering_phase = -1j * 2 * np.pi * ANTENNA_DIS * np.sin(np.deg2rad(ANGLE_DES)) / WAVE_LEN
Steering_des = np.exp(steering_phase * indices)


def steering_column(angle_deg: float) -> np.ndarray:
    """Steering vector (column) evaluated at angle_deg degrees."""
    phase = -1j * 2 * np.pi * ANTENNA_DIS * np.sin(np.deg2rad(angle_deg)) / WAVE_LEN
    return np.exp(phase * indices)[:, None]


w = cp.Variable(NUM_ANA, complex=True)
t = cp.Variable()  # proxy for the largest sidelobe level across all blocked angles

# Distortionless constraint: preserve amplitude/phase toward ANGLE_DES.
constraints = [cp.sum(cp.multiply(np.conj(Steering_des), w)) == 1]

# Enumerate each sidelobe direction and force its quadratic output ≤ t.
for angle in range(-90, theta_l + 1):
    vec = steering_column(angle)
    P_matrix = vec @ vec.conj().T
    constraints.append(cp.quad_form(w, P_matrix) <= t)
for angle in range(theta_u, 91):
    vec = steering_column(angle)
    P_matrix = vec @ vec.conj().T
    constraints.append(cp.quad_form(w, P_matrix) <= t)

# Worst-case design minimizes t such that every sidelobe angle obeys |wᴴa(θ)|² ≤ t.
objective = cp.Minimize(t)
problem = cp.Problem(objective, constraints)
problem.solve(solver=cp.CLARABEL)

angles = np.arange(-90, 91)
steering_vec_plot = np.exp(
    -1j * np.outer(np.sin(np.deg2rad(angles)), indices) * 2 * np.pi * ANTENNA_DIS / WAVE_LEN
)
# Convert the complex array response into a power pattern in dB.
pattern_db = 10 * np.log10(np.abs(np.conj(w.value) @ steering_vec_plot.T) ** 2 + 1e-12)

fig, ax = plt.subplots()
ax.plot(angles, pattern_db)
ax.set_title("Minimizing the Worst-case Sidelobe")
ax.grid(True)
ax.set_xlabel("Angle (degree)")
ax.set_ylabel("Angle Response (dB)")
ax.set_xlim([-90, 90])
fig.savefig("problem3a.png", dpi=300, bbox_inches="tight")
with open("problem3a.fig.pickle", "wb") as fig_file:
    pickle.dump(fig, fig_file)
plt.show()
plt.close(fig)
